### PR TITLE
Update to support wallet adapter v2

### PR DESCRIPTION
**Pull request by hardsetting**
Update to wallet adapter v2 and sponsored transactions

### CODE REVIEW

**Review by 0xmaayan — COMMENTED** [review]
ohhh ok I reviewed the other [PR](https://github.com/aptos-labs/petra-plugin-wallet-adapter/pull/15#pullrequestreview-1839425651) before this one - same question here then

I wonder if we need to hold the conversion layer in the plugin level or on the adapter. A proposal plan to support both v1 and v2 is to have the plugin hod a version?:"v1"|"v2" prop and the adapter to do the i/o conversion.

File: src/conversion.ts [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+}
+
+export function convertV2toV1<TV1, TV2 extends SerializableV2>(src: TV2, dst: DeserializableV1Class<TV1>) {
+  const serializedBytes = src.bcsToBytes();
+  const deserializerV1 = new BCS.Deserializer(serializedBytes);
+  return dst.deserialize(deserializerV1);
+}
+
+// old => new
+export function convertNetwork(networkInfo: NetworkInfo | null): Network {
```
Comment (0xmaayan) [inline src/conversion.ts:31]: the adapter holds a conversion layer with some same functionality - we might want to keep all conversions there
https://github.com/aptos-labs/aptos-wallet-adapter/blob/main/packages/wallet-adapter-core/src/conversion.ts

File: src/index.ts [hunk 2]
```diff
@@ -135,10 +199,10 @@ export class PetraWallet implements AdapterPlugin {
...
     try {
       const handleNetworkChange = async ({
-        name,
-        chainId,
-        url,
-      }: NetworkInfo): Promise<void> => {
+                                           name,
+                                           chainId,
+                                           url,
+                                         }: NetworkInfo): Promise<void> => {
```
Comment (0xmaayan) [inline src/index.ts:205]: thats a weird format lol

File: src/index.ts [hunk 3]
```diff
@@ -23,6 +27,7 @@ export const PetraWalletName = "Petra" as WalletName<"Petra">;
 
 export class PetraWallet implements AdapterPlugin {
   readonly name = PetraWalletName;
+  readonly version = "v2";
```
Comment (0xmaayan) [inline src/index.ts:30]: having it here will make the adapter to use this plugin as `v2` for everything - types/logic/etc. just making sure we are ok with it

File: src/index.ts (hunk 2 above)
Comment (hardsetting) [reply src/index.ts:205]: agreed 😆 
I removed the whole thing in #15, if it's cool we can leave it here and hopefully we merge both PRs together?

File: src/conversion.ts (hunk 1 above)
Comment (hardsetting) [reply src/conversion.ts:31]: See my comment https://github.com/aptos-labs/petra-plugin-wallet-adapter/pull/15#issuecomment-1906751125

**Comment by hardsetting** [comment]
> I wonder if we need to hold the conversion layer in the plugin level or on the adapter. A proposal plan to support both v1 and v2 is to have the plugin hod a version?:"v1"|"v2" prop and the adapter to do the i/o conversion.

See my comment https://github.com/aptos-labs/petra-plugin-wallet-adapter/pull/15#issuecomment-1906751125

**Pull request merged**